### PR TITLE
DOC: add section to Sharp Bits discussing implicit list conversions

### PR DESCRIPTION
**Pull request by jakevdp**
Related to #6833

Rendered version: https://jax--6835.org.readthedocs.build/en/6835/notebooks/Common_Gotchas_in_JAX.html#non-array-inputs

### CODE REVIEW

File: docs/notebooks/Common_Gotchas_in_JAX.ipynb [hunk 1]
```diff
@@ -568,6 +568,209 @@
     "Note also that, as the two behaviors described above are not inverses of each other, reverse-mode automatic differentiation (which turns index updates into index retrievals and vice versa) [will not preserve the semantics of out of bounds indexing](https://github.com/google/jax/issues/5760). Thus it may be a good idea to think of out-of-bounds indexing in JAX as a case of [undefined behavior](https://en.wikipedia.org/wiki/Undefined_behavior)."
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "LwB07Kx5sgHu"
+   },
+   "source": [
+    "## 🔪 Non-array inputs\n",
```
Comment (8bitmp3) [inline docs/notebooks/Common_Gotchas_in_JAX.ipynb:577]: Useful (but maybe not necessary?) suggestion:

- The JAX Sharp Bits docs is getting longer, so this important "sharp" section may be easy to skip when scanning, especially if you're a non-technical user - "attention is all your need"

```suggestion
    "## 🔪 Non-array inputs: NumPy vs. JAX\n",
```

File: docs/notebooks/Common_Gotchas_in_JAX.ipynb [hunk 2]
```diff
@@ -568,6 +568,209 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "QPliLUZztxJt"
+   },
+   "source": [
+    "This is a deliberate design choice, because passing lists or tuples to traced functions can lead to silent performance degradation that might otherwise be difficult to detect.\n",
+    "\n",
+    "For example, consider this the following permissive version of `jnp.sum` that allows list inputs:"
```
Comment (8bitmp3) [inline docs/notebooks/Common_Gotchas_in_JAX.ipynb:647]: Nit:

```suggestion
    "For example, consider the following permissive version of `jnp.sum` that allows list inputs:"
```

or maybe I misread this @jakevdp

**Review by 8bitmp3 — COMMENTED** [review]
This new section awesome.

I have a few small "non-so-sharp" bits:

- The JAX Sharp Bits notebook is getting longer, so this important "sharp" section may be easy to skip when scanning, especially if you're a non-technical user - "attention is all your need":

```diff
## 🔪 Non-array inputs
## 🔪 Non-array inputs: NumPy vs. JAX
```

- And a potential typo thing ("consider this the following" -> "consider the following"):

```
For example, consider the following permissive version of `jnp.sum` that allows list inputs:
```

Thanks for adding this awesome implicit list conversion section 👍

**Review by froystig — APPROVED** [review]

File: docs/notebooks/Common_Gotchas_in_JAX.ipynb (hunk 1 above)
Comment (jakevdp) [reply docs/notebooks/Common_Gotchas_in_JAX.ipynb]: Done, thanks!

File: docs/notebooks/Common_Gotchas_in_JAX.ipynb (hunk 2 above)
Comment (jakevdp) [reply docs/notebooks/Common_Gotchas_in_JAX.ipynb]: Done!

**Pull request merged**